File: cassandra-service-base/src/main/scala/com/github/sedovalx/cassandra/services/base/CassandraMapperGenericImpl.scala
package com.github.sedovalx.cassandra.services.base

import java.lang.reflect.ParameterizedType
import java.util.Optional
import java.util.concurrent.CompletableFuture

import com.datastax.driver.core.Statement
import com.datastax.driver.mapping.{Mapper, MappingManager}
import net.javacrumbs.futureconverter.java8guava.FutureConverter

import com.github.sedovalx.cassandra.services.utils.Java8Interop.{toJavaFunction, toOptional}

import scala.annotation.varargs

/**
  * Created by Alexander
  * on 26.12.2015.
  */
abstract class CassandraMapperGenericImpl[Entity](mappingManager: MappingManager) extends CassandraMapperGeneric[Entity] {

    protected val mapper: Mapper[Entity] = mappingManager.mapper(getDomainClass)

    override def saveQuery(entity: Entity): Statement = mapper.saveQuery(entity)

    @varargs
    override def save(entity: Entity, options: Mapper.Option*): Unit = mapper.save(entity, options:_*)

    @varargs
    override def saveAsync(entity: Entity, options: Mapper.Option*): CompletableFuture[Void] = {
        FutureConverter.toCompletableFuture(mapper.saveAsync(entity, options:_*))
    }

    override def deleteQuery(entity: Entity): Statement = mapper.deleteQuery(entity)

    @varargs
    override def delete(entity: Entity, options: Mapper.Option*) = mapper.delete(entity, options:_*)

    @varargs
    override def deleteAsync(entity: Entity, options: Mapper.Option*): CompletableFuture[Void] = {
        FutureConverter.toCompletableFuture(mapper.deleteAsync(entity, options:_*))
    }

    private def getDomainClass: Class[Entity] = {
        this.getClass.getGenericSuperclass.asInstanceOf[ParameterizedType].getActualTypeArguments.head.asInstanceOf[Class[Entity]]
    }

    protected def deleteAsyncWithMapper(args: Array[AnyRef]): CompletableFuture[Void] = {
        FutureConverter.toCompletableFuture(mapper.deleteAsync(args:_*))
    }

    @varargs
    protected def combineArgumentsToVarargs(varargs: Array[AnyRef], args: AnyRef*): Array[AnyRef] = {
        (args ++ varargs).toArray
    }

    @varargs
    protected def getOptionalEntity(objects: AnyRef*): Optional[Entity] = Optional.ofNullable(mapper.get(objects:_*))

    @varargs
    protected def getOptionalFuture(objects: AnyRef*): CompletableFuture[Optional[Entity]] = {
        FutureConverter.toCompletableFuture(mapper.getAsync(objects:_*)).thenApply(toJavaFunction(toOptional))
    }
}

File: cassandra-service-base/src/main/java/com/github/sedovalx/cassandra/services/base/CassandraMapperGenericImpl.java
package com.github.sedovalx.cassandra.services.base;

import com.datastax.driver.core.Statement;
import com.datastax.driver.mapping.Mapper;
import com.datastax.driver.mapping.MappingManager;
import net.javacrumbs.futureconverter.java8guava.FutureConverter;

import java.lang.reflect.ParameterizedType;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * Created by Alexander
 * on 22.03.2016.
 */
public abstract class CassandraMapperGenericImpl<Entity> implements CassandraMapperGeneric<Entity> {
    private final Mapper<Entity> mapper;

    protected Mapper<Entity> mapper() {
        return mapper;
    }

    public CassandraMapperGenericImpl(MappingManager mappingManager) {
        this.mapper = mappingManager.mapper(getDomainClass());
    }

    @Override
    public Statement saveQuery(Entity entity) {
        return mapper.saveQuery(entity);
    }

    @Override
    public void save(Entity entity, Mapper.Option... options) {
        mapper.save(entity, options);
    }

    @Override
    public CompletableFuture<Void> saveAsync(Entity entity, Mapper.Option... options) {
        return FutureConverter.toCompletableFuture(mapper.saveAsync(entity, options));
    }

    @Override
    public Statement deleteQuery(Entity entity) {
        return mapper.deleteQuery(entity);
    }

    @Override
    public void delete(Entity entity, Mapper.Option... options) {
        mapper.delete(entity, options);
    }

    @Override
    public CompletableFuture<Void> deleteAsync(Entity entity, Mapper.Option... options) {
        return FutureConverter.toCompletableFuture(mapper.deleteAsync(entity, options));
    }

    protected CompletableFuture<Void> deleteAsyncWithMapper(Object[] args) {
        return FutureConverter.toCompletableFuture(mapper.deleteAsync(args));
    }

    protected Object[] combineArgumentsToVarargs(Object[] varargs, Object... args) {
        Object[] array = new Object[varargs.length + args.length];
        System.arraycopy(varargs, 0, array, 0, varargs.length);
        System.arraycopy(args, varargs.length, array, varargs.length, varargs.length + args.length - varargs.length);
        return array;
    }

    protected Optional<Entity> getOptionalEntity(Object... args) {
        return Optional.ofNullable(mapper.get(args));
    }

    protected CompletableFuture<Optional<Entity>> getOptionalFuture(Object... args) {
        return FutureConverter.toCompletableFuture(mapper.getAsync(args)).thenApply(Optional::ofNullable);
    }

    private Class<Entity> getDomainClass(){
        return (Class<Entity>)((ParameterizedType)this.getClass().getGenericSuperclass()).getActualTypeArguments()[0];
    }
}
